solidity/contracts/VRF.sol

```diff
@@ -0,0 +1,252 @@
+
+  // Prime characteristic of the galois field over which the curve is defined
+  // github.com/ethereum/go-ethereum/blob/1636d957/crypto/bn256/cloudflare/constants.go#L23
+  // N.B.: The comments there are inaccurate. This is not actually curve BN256!!
+  // (It's Alt-BN128.)
+  uint256 constant fieldSize = 21888242871839275222246405745257275088696311157297823662689037894645226208583;
+
+  uint256 constant minusOne = fieldSize.sub(1);
+  uint256 constant multiplicativeGroupOrder = fieldSize.sub(1);
+  uint256 constant eulersCriterionPower = multiplicativeGroupOrder.div(2);
+  uint256 constant sqrtPower = fieldSize.add(1).div(4); // Works since fieldSize % 4 = 3
+
+  // github.com/ethereum/go-ethereum/blob/1636d957/crypto/bn256/cloudflare/curve.go#L15
+  uint256[2] private generator = [uint256(1), uint256(2)];
+  // Number of GF(fieldSize) points on the elliptic curve
+  // github.com/ethereum/go-ethereum/blob/1636d957/crypto/bn256/cloudflare/constants.go
+  uint256 constant groupOrder =  21888242871839275222246405745257275088548364400416034343698204186575808495617;
+
+  // Bits which can be used in the representation of a number less than
+  // groupOrder or fieldSize (i.e., the lower 254 bits.)
+  uint256 constant orderMask = (~uint256(0)) >> 2;
+
+  uint256 constant wordLengthBytes = 0x20;
+
+  // (_base**_exponent) % _modulus
+  // Cribbed from https://medium.com/@rbkhmrcr/precompiles-solidity-e5d29bd428c4
+  function bigModExp(uint256 _base, uint256 _exponent, uint256 _modulus)
+    public view returns (uint256 exponentiation) {
+    uint256 callResult;
+    uint256[6] memory bigModExpContractInputs;
+    bigModExpContractInputs[0] = wordLengthBytes;  // Length of _base
+    bigModExpContractInputs[1] = wordLengthBytes;  // Length of _exponent
+    bigModExpContractInputs[2] = wordLengthBytes;  // Length of _modulus
+    bigModExpContractInputs[3] = _base;
+    bigModExpContractInputs[4] = _exponent;
+    bigModExpContractInputs[5] = _modulus;
+    uint256[1] memory output;
+    assembly {
+      callResult :=
+        staticcall(                          // Let bigmodexp use arbitrary gas
+                   0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,
+                   0x05,                     // Bigmodexp contract address
+                   bigModExpContractInputs,
+                   0xc0,                     // Length of input segment
+                   output,
+                   0x20)                     // Length of output segment
+      }
+    if (callResult == 0) {revert("bigModExp failure!");}
+    return output[0];
+  }
+
+  // _scalar*p, in the curve group. Reverts if p not on curve.
+  function scalarMul(uint256[2] memory _p, uint256 _scalar)
+    public view returns (uint256[2] memory xyOutput) {
+    uint256 callResult;
+    uint256[3] memory inputs;
+    inputs[0] = _p[0];
+    inputs[1] = _p[1];
+    inputs[2] = _scalar;
+    assembly {
+      callResult := // See comments on bigModExp assembly.
```

Comment:
This is actually a place where I think more comments would help. A little duplication is worth it, the purpose of the values isn't going to change and it removes the need to jump around the code when reading.

One day we'll get constants in assembly.🤞